<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="What the package does (one paragraph).">
<title>scX • scX</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="deps/_Roboto-0.4.1/font.css" rel="stylesheet">
<link href="deps/_JetBrains%20Mono-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="scX">
<meta property="og:description" content="What the package does (one paragraph).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">scX</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/install.html">Install</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/get_started.html">Quick Start</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tabs">Tabs</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tabs">
    <a class="dropdown-item" href="articles/Summary.html">Summary</a>
    <a class="dropdown-item" href="articles/Markers.html">Markers</a>
    <a class="dropdown-item" href="articles/Gene_Expression.html">Gene Expression</a>
    <a class="dropdown-item" href="articles/Differential_Expression.html">Differential Gene Expression</a>
    <a class="dropdown-item" href="articles/Exploratory_Data_Analysis.html">Explotarory Data Analysis</a>
    <a class="dropdown-item" href="articles/Visual_Tools.html">Visual Tools</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/FAQ.html">FAQ</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.leloir.org.ar/biologia-de-sistemas-integrativa?area=bioinformatica-y-biologia-computacional">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/tvegawaichman/scX">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header"><h1 align="center">
scXplorer package
</h1></div>
<div align="justify">
<p><img align="left" width="25%" src="reference/figures/scXplorer-03.png" title="@irimiodo"> scXplorer is an R package that enables interactive visualization and analysis of single-cell experiment data by creating a Shiny-based application. With scXplorer all aspects of single-cell data can be explored using a large number of different types of plots, such as scatter plots, heatmaps, boxplots, dot and violins plots, etc. All the information associated with cells can be displayed in a customized way: both numerical variables such as logcounts or pseudotime values, and categorical variables such as cell types or sample. One of the main hallmarks of scXplorer is the possibility to plot the main embeddings used for single cell - UMAP, tSNE and PCA - both in 2D and 3D in an interactive way. Thus, embeddings can be rotated, translated and zoomed. But scXplorer is not only a visualization tool, it also allows you to perform different types of analysis on single cell data, such as finding the markers of a cell type or determining the differential genes between two different conditions.</p>
</div>
<hr>
<table width="100%" style="border:0px solid white; width:100%;" class="table"><tr style="border: 0px;">
<td width="33%" style="border:0px; width:33.33%">
<h2 align="left">
Table of Contents
</h2>
<ul>
<li><a href="#install">Install</a></li>
<li>
<a href="#quick-start-guide">Quick Start Guide</a>
<ul>
<li><a href="#summary">Summary</a></li>
<li><a href="#markers">Markers</a></li>
<li><a href="#gene-expression">Gene expression</a></li>
<li><a href="#differential-expression">Differential expression</a></li>
<li><nobr><a href="#exploratory-data-analysis">Exploratory Data Analysis</a> </nobr></li>
<li><a href="#visual-tools">Visual Tools</a></li>
</ul>
</li>
<li><a href="#faqs">FAQs</a></li>
</ul>
</td>
<td width="66.66%" style="border:0px; width:66.66%">
<p><img src="reference/figures/presentation_t2.gif" width="100%"></p>
</td>
</tr></table>
<div class="section level2">
<h2 id="install">Install<a class="anchor" aria-label="anchor" href="#install"></a>
</h2>
<p>scXplorer requires certain packages to be installed. If the user does not have these packages installed, they can execute the following code:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">"SingleCellExperiment"</span>,<span class="st">"scran"</span>,<span class="st">"scatter"</span>,<span class="st">"ComplexHeatmap"</span>)</span></code></pre></div>
<p>scXplorer can be installed from Github as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"tvegawaichman/scXplorer"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start-guide">Quick Start Guide<a class="anchor" aria-label="anchor" href="#quick-start-guide"></a>
</h2>
<div class="section level3">
<h3 id="loading-datasets">Loading Datasets<a class="anchor" aria-label="anchor" href="#loading-datasets"></a>
</h3>
<p>To show the different capabilities of scXplorer, we will use single cell data related to the oligodendrocyte developmental lineage (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5221728/" class="external-link">Marques et al. 2016</a>). In this dataset we have cells from 12 regions of the central nervous system of juvenile and adult mice and 10 distinct cell populations have been identified. This package includes a modified version of this dataset in which the original cells have been subsampled and a pseudotime has been calculated to demonstrate how scXplorer can represent numerical attributes.</p>
<p>In this example we will use the ‘inferred_cell_type’ covariate to explore differential expression patterns (partitionVars parameter). The rest of the metadata will be used for visualization purposes (metadataVars parameter). While partitionVars should be categorical variables, metadataVars can be discrete or continuous covariates</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scXplorer</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#SCE example data</span></span>
<span><span class="va">sce</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span></span>
<span><span class="co"># DataFrame with 3 rows and 8 columns</span></span>
<span><span class="co">#                                   title     source_name         age inferred_cell_type                    sex      strain   treatment pseudotime</span></span>
<span><span class="co">#                             &lt;character&gt;     &lt;character&gt; &lt;character&gt;        &lt;character&gt;            &lt;character&gt; &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#   C1-1771017-030-C09 C1-1771017-030-C09       cortex S1         p22                OPC pooled male and female  PDGFRa-GFP          No    23.1662</span></span>
<span><span class="co">#   C1-1771017-028-G05 C1-1771017-028-G05 hippocampus CA1         p22                OPC                      F  PDGFRa-GFP          No    21.7566</span></span>
<span><span class="co">#   C1-1771052-132-B02 C1-1771052-132-B02 corpus callosum         p69                OPC                      M         CD1          No    23.3207</span></span></code></pre></div>
<p align="justify">
</p>
<p>The scXplorer app can be created and launched with only two functions. <strong><em>createSCEobject()</em></strong> creates the single cell experiment object that will be used within the application. This function performs a set of preprocessing steps required for the various scXplorer functions. Among these steps we can mention:</p>
<ol style="list-style-type: decimal">
<li><p>Calculation of quality control metrics.</p></li>
<li><p>If no partition was declared in the “partitionVars” parameter, a preliminary clustering will be calculated with the scran package function <em>quickCluster()</em>.</p></li>
<li><p>Normalization of the gene expression matrix.</p></li>
<li><p>Determination of the most variable genes (HVG).</p></li>
<li><p>Calculation of different dimensionality reductions: PCA, tSNE, and UMAP.</p></li>
</ol>
<p>If the input dataset already has these characteristics, some of these steps can be avoided. For example, if the dataset already has a precomputed PCA, scXplorer will not recalculate it but will calculate the dimensionality reductions that are not already calculated in the dataset.</p>
<p>On the other hand, different options for the calculation of the marker genes of the different clusters can be determined. This can be done within the parameter “paramFindMarkers” which expects a list of parameters taken by scran’s <em>FindMarkers()</em> function.</p>
<p>Finally with the <strong><em>launch_scXplorer()</em></strong> function the application is deployed.</p>

<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scXplorer</span><span class="op">)</span></span>
<span><span class="co"># Creating SCE object</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="st">"inferred_cell_type"</span>, </span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"source_name"</span>, <span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"strain"</span>, <span class="st">"treatment"</span>, <span class="st">"pseudotime"</span><span class="op">)</span>,</span>
<span>                        descriptionText <span class="op">=</span> <span class="st">"Quick Start Guide"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/launch_scXplorer.html">launch_scXplorer</a></span><span class="op">(</span><span class="va">cseo</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p align="justify">
</p>
<p>scXplorer displays a summary of the main descriptive information of the dataset: number of cells and genes, mean number of genes detected per cell, partions, experiments and average library size.</p>
In the summary section, you can explore through graphical visualization the coventional quality control metrics: this is the relationship between the number of features and the count numbers. In this plot you can distinguish the cells coming from the different partitions of the dataset.

<details><summary><h5>
View more
</h5>
</summary><p><img src="reference/figures/summary.gif" width="100%"></p>
</details>
</div>
<div class="section level2">
<h2 id="markers">Markers<a class="anchor" aria-label="anchor" href="#markers"></a>
</h2>
<p align="justify">
This section allows to find and interact with the markers of each different clusterization. In “Markers” section there are two types of analysis. On the one hand, in “Cluster markers” clicking on a cell displays a table with the marker genes of the cluster to which that cell belongs. On the other hand, in “Find new markers” you can select a group of cells in the embedding and scXplorer will calculate their marker genes.
</p>
<details><summary><h5>
View more
</h5>
</summary><div class="section level3">
<h3 id="clusters-markers">Clusters markers<a class="anchor" aria-label="anchor" href="#clusters-markers"></a>
</h3>
<p align="justify">
</p>
<p>This section allows you to find the marker genes for the partition defined in the single-cell object, typically cell types or cell states. Clicking on one of the cells in the embedding will display a table of marker genes for the partition to which that cell belongs. For each of the markers different metrics such as boxcor, robustness and FDR are displayed. This table can be downloaded in various formats, such as .csv, .xlsx .pdf, or you can copy it to the clipboard.</p>
By clicking on a marker in the table you can see its expression profile across the entire dataset in the embedding. In addition violin and spikeplots are displayed at the bottom.

<p><img src="reference/figures/cluster_markers.gif" width="100%"></p>
</div>
<div class="section level3">
<h3 id="find-new-markers">Find new markers<a class="anchor" aria-label="anchor" href="#find-new-markers"></a>
</h3>
<p align="justify">
</p>
<p>Here you can select with the box or lasso tool a set of cells in the embedding and scXplorer will calculate the marker genes. You can download not only the marker table but also the selected cell list.</p>
As in the previous section, if you click on one of the markers you can see its expression along the dataset with violin and spikeplots.

<img src="reference/figures/new_markers.gif" width="100%">
</div>
</details>
</div>
<div class="section level2">
<h2 id="gene-expression">Gene Expression<a class="anchor" aria-label="anchor" href="#gene-expression"></a>
</h2>
<p align="justify">
In “Gene Expression” you can explore different aspects of the expression of one or more genes of interest. Determine how expression changes according to different categorical and continuous variables, as well as analyse co-detection between pairs of genes.
</p>
<details><summary><h5>
View more
</h5>
</summary><div class="section level3">
<h3 id="categories">Categories<a class="anchor" aria-label="anchor" href="#categories"></a>
</h3>
<p align="justify">
</p>
<p>In Settings you can select one or more genes or upload a file with a list of genes. The average expression of the genes of interest can be viewed in the different embeddings available, with the possibility to colour according to the different SCE partitions to compare gene expression with different cell types or conditions present in the metadata.</p>
A wide variety of plots are available to analyse the expression of the genes of interest in the different categories. Heatmaps allow normalisation of expression by gene, clustering by row and column and grouping of cells by condition. Similarly, dotplots allow normalisation of expression and group genes with similar expression pattern.

<p><img src="reference/figures/categories.gif" width="100%"></p>
</div>
<div class="section level3">
<h3 id="fields">Fields<a class="anchor" aria-label="anchor" href="#fields"></a>
</h3>
<p align="justify">
Fields allows you to analyse the expression of a set of genes in relation to numeric variables present in your dataset, such as the number of counts or pseudotime value, if present in the metadata of the sce object. Below the embedding, a line plot of the average expression of the genes of interest as a function of the chosen variables and a spikeplot are displayed. Furthermore, you can find heatmaps sorted by the chosen numerical variable that can be divided according to some categorical variable, and multiline plots showing the comparison of the expression profile of the genes of interest along the field.
</p>
<p><img src="reference/figures/ge_fields.gif" width="100%"></p>
</div>
<div class="section level3">
<h3 id="co-expression">Co-expression<a class="anchor" aria-label="anchor" href="#co-expression"></a>
</h3>
<p align="justify">
In Co-expression section you can analyse the co-appearance of pairs of genes, determine the number and percentage of cells in which each gene is expressed separately and together. You can also view this information graphically in the embedding. In addition, the co-expression of these genes in the different conditions of any of the partitions in the dataset can be analysed by a co-detection matrix.
</p>
<p><img src="reference/figures/coexpression.gif" width="100%"></p>

</div>
</details>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p align="justify">
</p>
<p>In this section, the differentially expressed genes between two clusters can be obtained. The algorithm used is <code>findMarkers()</code> from the scran package and the test type can be defined in the <code><a href="reference/createSCEobject.html">createSCEobject()</a></code> function (see “Summary” section). Different significance levels can be chosen interactively for both the logFC and the FDR. Once the clusters have been chosen, a table will be displayed with the list of differentially expressed genes accompanied by the logFC and the FDR obtained for each of them. This table can be downloaded in different formats: csv, pdf, xlsx.</p>
The main figure in this section is a VolcanoPlot in which genes that are down and up expressed are coloured in blue and red, respectively. On the other hand, you can also display ViolinPlots, Spikeplots, Heatmaps and Dotplots of the differentially expressed genes.

<details><summary><h5>
View more
</h5>
</summary><p><img src="reference/figures/differential_expression.gif" width="100%"></p>
</details>
</div>
<div class="section level2">
<h2 id="exploratory-data-analysis">Exploratory Data Analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h2>
<p>In <strong>Exploratory Data Analysis</strong> section you will be able to understand the relationship between different features contained as metadata in your SCE object.</p>
<details><summary><h5>
View more
</h5>
</summary><div class="section level3">
<h3 id="categories-1">Categories<a class="anchor" aria-label="anchor" href="#categories-1"></a>
</h3>
<p align="justify">
Here you can observe the proportion of cells belonging to the different levels of a categorical variable presented in the metadata and disaggregate these proportions according to the levels of another categorical variable. All this information is displayed in the form of a barplot. In the subsection “Matrix” a confusion matrix between the two selected features can be plotted with the option to display the Jaccard index for each of the grid cells. In addition, the Rand index is displayed, which is a global measure of the similarity between the two clusterings.
</p>
<p><img src="reference/figures/exploratory_categories.gif" width="100%"></p>
</div>
<div class="section level3">
<h3 id="fields-1">Fields<a class="anchor" aria-label="anchor" href="#fields-1"></a>
</h3>
<p>In a similar way to the previous subsection, in “Field” you can explore how the value of one or more numerical variables changes as a function of another variable, either numerical or categorical. You can make different types of plots such as: Distribution Plots, Heatmaps, Dotplots and StackedViolins.</p>
<p><img src="reference/figures/fields.gif" width="100%"></p>

</div>
</details>
</div>
<div class="section level2">
<h2 id="visual-tools">Visual Tools<a class="anchor" aria-label="anchor" href="#visual-tools"></a>
</h2>
<p>In the “Visual Tools” section, different plots can be obtained to explore in more depth different aspects of the single-cell experiment data, such as how the expression of a given set of genes varies at different levels of a feature or to recognise a set of cells of interest within an embedding.</p>
<details><summary><h5>
View more
</h5>
</summary><div class="section level3">
<h3 id="violin-by-partition">Violin by Partition<a class="anchor" aria-label="anchor" href="#violin-by-partition"></a>
</h3>
<p>By selecting or uploading a set of genes of interest, a set of Violin plots can be obtained and download in .pdf format, allowing to facet the plots by differents categorical co-variables.</p>
<p><img src="reference/figures/violins.gif" width="100%"></p>
</div>
<div class="section level3">
<h3 id="multiplots">Multiplots<a class="anchor" aria-label="anchor" href="#multiplots"></a>
</h3>
<p>Multiplots allows you to explore how different variables change across cells in an embedding of your choice, such as the expression of a given set of genes, the partitions of a categorical variable or the value of a continuous variable.</p>
<p><img src="reference/figures/multiplots.gif" width="100%"></p>

</div>
</details>
</div>
<div class="section level2">
<h2 id="faqs">FAQs<a class="anchor" aria-label="anchor" href="#faqs"></a>
</h2>
<details><summary><h3>
1 - What type of input does scXplorer accept?
</h3>
</summary><blockquote>
</blockquote>
</details><details><summary><h3>
2 - Is there a maximum input size that scXplorer can receive?
</h3>
</summary><blockquote>
</blockquote>
</details><details><summary><h3>
3 - How can I deploy my scXplorer app online?
</h3>
</summary><blockquote>
</blockquote>
</details>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scX</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Tomás Vega Waichman <br><small class="roles"> Author, maintainer </small>  </li>
<li>Maria Luz Vercesi <br><small class="roles"> Author </small>  </li>
<li>Ariel Agustín Berardino <br><small class="roles"> Author </small>  </li>
<li>Maximiliano Sebastian Beckel <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Tomás Vega Waichman, Maria Luz Vercesi, Ariel Agustín Berardino, Maximiliano Sebastian Beckel.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
