<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scX">
<title>FAQ • scX</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.5/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.5/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FAQ">
<meta property="og:description" content="scX">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/get_started.html">Quick Start</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tabs">Tabs</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tabs">
    <a class="dropdown-item" href="../articles/Summary.html">Summary</a>
    <a class="dropdown-item" href="../articles/Markers.html">Markers</a>
    <a class="dropdown-item" href="../articles/Gene_Expression.html">Gene Expression</a>
    <a class="dropdown-item" href="../articles/Differential_Expression.html">Differential Gene Expression</a>
    <a class="dropdown-item" href="../articles/Exploratory_Data_Analysis.html">Explotarory Data Analysis</a>
    <a class="dropdown-item" href="../articles/Visual_Tools.html">Visual Tools</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.leloir.org.ar/biologia-de-sistemas-integrativa?area=bioinformatica-y-biologia-computacional">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/chernolabs/scX">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>FAQ</h1>
            
      
      
      <div class="d-none name"><code>FAQ.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="what-type-of-input-does-scx-accept">1. What type of input does scX accept?<a class="anchor" aria-label="anchor" href="#what-type-of-input-does-scx-accept"></a>
</h2>
<p>scX is a “plug and play” tool that pre-processes the data using only
the raw count matrix, allowing the dataset to be explored in depth.
Nonetheless, it is advisable to input an <strong>SCE object</strong>
when using the <code><a href="../reference/createSCEobject.html">createSCEobject()</a></code> function. The function,
however, also accepts the <strong>raw count matrix</strong> (with genes
as rows and cells as columns) or a <strong>Seurat object</strong>. In
the latter two cases, it automatically transforms the inputs into an SCE
object.</p>
<p>If you supply cell metadata, it must be contained in a dataframe with
the cell names as rownames. Typical examples of metadata are cell type,
sampleid, percentage of mitochondrial reads.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Input = SingleCellExperiment object (sce)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Input = Seurat object (seu)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">seu</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively, Input = combine a matrix (mat) from one of these classes:</span></span>
<span><span class="co"># c("dgCMatrix", "Matrix", "matrix") and metadata as data.frame (df)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">df</span><span class="op">)</span></span></code></pre></div>
<p>scX is an extremely valuable tool for visualising and analysing
single-cell or single-nuclei RNA sequencing data. To maximise its
potential, it is important to have a pre-processed object, i.e. one that
has already undergone QC filtering for genes and cells, as well as the
appropriate normalisation for the type of data. If you have already
computed partitions, such as a Louvain-Leiden partition to identify cell
types or states, you can pass them as input to be visualised in PCA,
UMAP, TSNE and obtain markers, DEGs between the different clusters, and
so on. In that case use the <code>partitionVars</code> parameter. To
visualize any additional information in plots, it must be passed through
the <code>metadataVars</code> parameter.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudotime"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="how-to-pass-a-normalized-count-matrix-as-input-of-createsceobject">How to pass a normalized count matrix as input of
<code>createSCEobject()</code>?<a class="anchor" aria-label="anchor" href="#how-to-pass-a-normalized-count-matrix-as-input-of-createsceobject"></a>
</h3>
<p>If you possess the normalized count matrix rather than the raw count
matrix, you may transform it into an SCE object before providing it to
<code><a href="../reference/createSCEobject.html">createSCEobject()</a></code></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logcounts<span class="op">=</span><span class="va">norm.mat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="is-there-a-maximum-input-size-that-scx-can-receive">2. Is there a maximum input size that <em>scX</em> can receive?<a class="anchor" aria-label="anchor" href="#is-there-a-maximum-input-size-that-scx-can-receive"></a>
</h2>
<p>The limitations of scX are solely dependent on the PC’s hardware,
such as the RAM and CPU. A ‘standard’ PC (16gb, i5-8400) can comfortably
manage medium-sized datasets of up to 30k cells. However, to ensure
smooth application and visualization plots for datasets exceeding 50k
cells, the <code><a href="../reference/createSCEobject.html">createSCEobject()</a></code> function subsamples the SCE
object with 50k random cells while storing the cell names in the
<code>CELLS2KEEP</code> output.</p>
<ul>
<li>If you have a cluster with only a few cells that you wish to retain,
you can pass their names to the <code>cells2keep</code> parameter.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">c2k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">sce</span><span class="op">$</span><span class="va">inferred_cell_type</span><span class="op">==</span><span class="st">"COP"</span><span class="op">]</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        cells2keep <span class="op">=</span> <span class="va">c2k</span><span class="op">)</span> <span class="co"># Keeping COP cells in case of subsampling data for visualization</span></span></code></pre></div>
<ul>
<li>If you need to utilize over 50k cells and possess a 200k cell SCE
object, implementing a stratified subsampling to preserve half the cells
in each cluster is a viable option. This will result in the app
exhibiting 100k cells during visualization.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co"># sce: SCE object of a dataset with 200k cells</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cellnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">c2k</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">as.data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">inferred_cell_type</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">sample_frac</a></span><span class="op">(</span><span class="fl">0.5</span>,replace <span class="op">=</span>  <span class="cn">F</span><span class="op">)</span> <span class="co"># sample half of the cells in every cluster</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudotime"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change the output of the subsampling function to be able to use 100k cells:</span></span>
<span><span class="va">cseo</span><span class="op">$</span><span class="va">CELLS2KEEP</span> <span class="op">&lt;-</span> <span class="va">c2k</span><span class="op">$</span><span class="va">cellnames</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>If you are willing to attempt to utilize all cells within the
application with the potential drawback of running low on memory or
experiencing extended waiting periods between the plots, you can modify
the parameter to <code>all</code> prior to launching
<code><a href="../reference/launch_scX.html">launch_scX()</a></code>.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudotime"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change the output of the subsampling function:</span></span>
<span><span class="va">cseo</span><span class="op">$</span><span class="va">CELLS2KEEP</span> <span class="op">&lt;-</span> <span class="st">'all'</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span><span class="op">)</span></span></code></pre></div>
<p><strong>Important note:</strong> All the calculations in the
preprocessing stage apply to every cell within the dataset, regardless
of the subsampling provided by the <code>cells2keep</code> parameter for
visualization purposes.</p>
</div>
<div class="section level2">
<h2 id="how-to-deploy-a-shiny-app-in-shiny-io">3. How to deploy a shiny app in shiny.io?<a class="anchor" aria-label="anchor" href="#how-to-deploy-a-shiny-app-in-shiny-io"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Save the SCE object in <code>/data</code> folder</li>
<li>Create an app.R script:</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://chernolabs.github.io/scX/">scX</a></span><span class="op">)</span></span>
<span><span class="co"># Loading SCE object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">'data/data.Rdata'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span>,point.size <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Then run:</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/rsconnect" class="external-link">rsconnect</a></span><span class="op">)</span></span>
<span><span class="fu">rsconnect</span><span class="fu">::</span><span class="fu">deployApp</span><span class="op">(</span><span class="st">'path/to/your/app.R'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="how-to-use-scanpy--h5ad-or--loom-objets">4. How to use Scanpy (.h5ad) or .loom objets?<a class="anchor" aria-label="anchor" href="#how-to-use-scanpy--h5ad-or--loom-objets"></a>
</h2>
<p>For this it’s necessary to convert them to a SingleCellExperiment or
Seurat objects. In this tutorial we will explain how to convert them as
SingleCellExperiment using the <a href="https://bioconductor.org/packages/release/bioc/html/zellkonverter.html" class="external-link">zellkonverter
package</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">readH5AD</span><span class="op">(</span><span class="va">file</span>,</span>
<span>        reader <span class="op">=</span>  <span class="st">"R"</span></span>
<span><span class="co">#Then run the createSCEobject as in 1)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>From .loom, using the <a href="https://github.com/cellgeni/sceasy" class="external-link">sceasy package</a></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sceasy</span><span class="fu">::</span><span class="fu">convertFormat</span><span class="op">(</span><span class="st">'filename.loom'</span>, from<span class="op">=</span><span class="st">"loom"</span>, to<span class="op">=</span><span class="st">"sce"</span>,</span>
<span>                       outFile<span class="op">=</span><span class="st">'filename.rds'</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'filename.rds'</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/tvegawaichman" class="external-link">Tomás Vega Waichman</a>, <a href="https://github.com/mluzvercesi" class="external-link">María Luz Vercesi</a>, <a href="https://github.com/ariberar" class="external-link">Ariel A. Berardino</a>, <a href="https://github.com/msbeckel" class="external-link">Maximiliano S. Beckel</a>, Chernomoretz Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
