<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scX">
<title>FAQ • scX</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.8/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FAQ">
<meta property="og:description" content="scX">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/get_started.html">Quick Start</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tabs">Tabs</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tabs">
    <a class="dropdown-item" href="../articles/Summary.html">Summary</a>
    <a class="dropdown-item" href="../articles/Exploratory_Data_Analysis.html">Explotarory Data Analysis</a>
    <a class="dropdown-item" href="../articles/Markers.html">Markers</a>
    <a class="dropdown-item" href="../articles/Gene_Expression.html">Gene Expression</a>
    <a class="dropdown-item" href="../articles/Differential_Expression.html">Differential Gene Expression</a>
    <a class="dropdown-item" href="../articles/Visual_Tools.html">Visual Tools</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.leloir.org.ar/biologia-de-sistemas-integrativa?area=bioinformatica-y-biologia-computacional">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/chernolabs/scX">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>FAQ</h1>
            
      
      
      <div class="d-none name"><code>FAQ.Rmd</code></div>
    </div>

    
    
<style type="text/css">
/* Define a margin before h2 element */
h2  {
  margin-top: 1em;
}

/* Define a margin before every first h4 elements */
h4:first-of-type {
  margin-top: 1em;
}
</style>
<div class="section level2">
<h2 id="what-type-of-input-does-scx-accept">1. What type of input does scX accept?<a class="anchor" aria-label="anchor" href="#what-type-of-input-does-scx-accept"></a>
</h2>
<p>scX is a “plug and play” tool that pre-processes the data using only
the raw count matrix, allowing the dataset to be explored in depth.
Nonetheless, it is advisable to input an <strong>SCE object</strong>
when using the <code><a href="../reference/createSCEobject.html">createSCEobject()</a></code> function. The function,
however, also accepts the <strong>raw count matrix</strong> (with genes
as rows and cells as columns) or a <strong>Seurat object</strong>. In
the latter two cases, it automatically transforms the inputs into an SCE
object.</p>
<p>If you supply cell metadata, it must be contained in a dataframe with
the cell names as rownames. Typical examples of metadata are cell type,
sampleid, percentage of mitochondrial reads.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Input = SingleCellExperiment object (sce)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Input = Seurat object (seu)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">seu</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alternatively, Input = combine a matrix (mat) from one of these classes:</span></span>
<span><span class="co"># c("dgCMatrix", "Matrix", "matrix") and metadata as data.frame (df)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">df</span><span class="op">)</span></span></code></pre></div>
<p>scX is an extremely valuable tool for visualising and analysing
single-cell or single-nuclei RNA sequencing data. To maximise its
potential, it is important to have a pre-processed object, i.e. one that
has already undergone QC filtering for genes and cells, as well as the
appropriate normalisation for the type of data. If you have already
computed partitions, such as a Louvain-Leiden partition to identify cell
types or states, you can pass them as input to be visualised in PCA,
UMAP, TSNE and obtain markers, DEGs between the different clusters, and
so on. In that case use the <code>partitionVars</code> parameter. To
visualize any additional information in plots, it must be passed through
the <code>metadataVars</code> parameter.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">scXample</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudotime"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="how-to-pass-a-normalized-count-matrix-as-input-of-createsceobject">1.1 How to pass a normalized count matrix as input of
<code>createSCEobject()</code>?<a class="anchor" aria-label="anchor" href="#how-to-pass-a-normalized-count-matrix-as-input-of-createsceobject"></a>
</h4>
<p>If you possess the normalized count matrix rather than the raw count
matrix, you may transform it into an SCE object before providing it to
<code><a href="../reference/createSCEobject.html">createSCEobject()</a></code></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logcounts<span class="op">=</span><span class="va">norm.mat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="how-to-use-scanpy--h5ad-or--loom-objets">1.2 How to use Scanpy (.h5ad) or .loom objets?<a class="anchor" aria-label="anchor" href="#how-to-use-scanpy--h5ad-or--loom-objets"></a>
</h4>
<p>For this it’s necessary to convert them to a SingleCellExperiment or
Seurat objects. In this tutorial we will explain how to convert them as
SingleCellExperiment using the <a href="https://bioconductor.org/packages/release/bioc/html/zellkonverter.html" class="external-link">zellkonverter
package</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">readH5AD</span><span class="op">(</span><span class="va">file</span>,</span>
<span>        reader <span class="op">=</span>  <span class="st">"R"</span></span>
<span><span class="co">#Then run the createSCEobject as in 1)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>From .loom, using the <a href="https://github.com/cellgeni/sceasy" class="external-link">sceasy package</a></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sceasy</span><span class="fu">::</span><span class="fu">convertFormat</span><span class="op">(</span><span class="st">'filename.loom'</span>, from<span class="op">=</span><span class="st">"loom"</span>, to<span class="op">=</span><span class="st">"sce"</span>,</span>
<span>                       outFile<span class="op">=</span><span class="st">'filename.rds'</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'filename.rds'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="is-there-a-maximum-size-limit-for-input-objects-that-scx-is-capable-of-processing">2. Is there a maximum size limit for input objects that <em>scX</em>
is capable of processing?<a class="anchor" aria-label="anchor" href="#is-there-a-maximum-size-limit-for-input-objects-that-scx-is-capable-of-processing"></a>
</h2>
<p>The constraints of scX are determined by the hardware of the
computer, including the RAM and CPU. A typical PC (16GB, i5-8400) can
efficiently handle datasets of medium size, containing up to 30k cells.
Be aware that the number of partitions and clusters will also affect the
processing time of a dataset.</p>
<p>To ensure smooth application and visualisation plots for datasets
with more than 50k cells, the <code><a href="../reference/createSCEobject.html">createSCEobject()</a></code> function
subsamples the SCE object with 50k random cells, storing the cell names
in the <code>CELLS2KEEP</code> output.</p>
<ul>
<li>If you have a cluster with only a few cells that you wish to retain,
you can pass their names to the <code>cells2keep</code> parameter.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">c2k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scXample</span><span class="op">)</span><span class="op">[</span><span class="va">scXample</span><span class="op">$</span><span class="va">inferred_cell_type</span><span class="op">==</span><span class="st">"COP"</span><span class="op">]</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">scXample</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        cells2keep <span class="op">=</span> <span class="va">c2k</span><span class="op">)</span> <span class="co"># Keeping COP cells in case of subsampling data for visualization</span></span></code></pre></div>
<ul>
<li>If you need to utilize over 50k cells and possess a 200k cell SCE
object, implementing a stratified subsampling to preserve half the cells
in each cluster is a viable option. This will result in the app
exhibiting 100k cells during visualization.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co"># sce: SCE object of a dataset with 200k cells</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cellnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">c2k</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">as.data.frame</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">inferred_cell_type</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">sample_frac</a></span><span class="op">(</span><span class="fl">0.5</span>,replace <span class="op">=</span>  <span class="cn">F</span><span class="op">)</span> <span class="co"># sample half of the cells in every cluster</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudotime"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change the output of the subsampling function to be able to use 100k cells:</span></span>
<span><span class="va">cseo</span><span class="op">$</span><span class="va">CELLS2KEEP</span> <span class="op">&lt;-</span> <span class="va">c2k</span><span class="op">$</span><span class="va">cellnames</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>If you are willing to attempt to utilize all cells within the
application with the potential drawback of running low on memory or
experiencing extended waiting periods between the plots, you can modify
the parameter to <code>all</code> prior to launching
<code><a href="../reference/launch_scX.html">launch_scX()</a></code>.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                        partitionVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"inferred_cell_type"</span><span class="op">)</span>, </span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pseudotime"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change the output of the subsampling function:</span></span>
<span><span class="va">cseo</span><span class="op">$</span><span class="va">CELLS2KEEP</span> <span class="op">&lt;-</span> <span class="st">'all'</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span><span class="op">)</span></span></code></pre></div>
<p><strong>Important note:</strong> All the calculations in the
preprocessing stage apply to every cell within the dataset, regardless
of the subsampling provided by the <code>cells2keep</code> parameter for
visualization purposes.</p>
</div>
<div class="section level2">
<h2 id="how-to-deploy-a-shiny-app-in-shinyapps-io">3. How to deploy a shiny app in shinyapps.io?<a class="anchor" aria-label="anchor" href="#how-to-deploy-a-shiny-app-in-shinyapps-io"></a>
</h2>
<p>One way to deploy an app online is to use the
<strong>shinyapps</strong> service. The first step is to create an
account at <a href="https://www.shinyapps.io/" class="external-link">https://www.shinyapps.io/</a>. A user
guide can be found <a href="https://docs.posit.co/shinyapps.io/index.html" class="external-link">here</a>.</p>
<ol style="list-style-type: decimal">
<li>Save the SCE object in <code>/data</code> folder</li>
<li>Create an app.R script:</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://chernolabs.github.io/scX/">scX</a></span><span class="op">)</span></span>
<span><span class="co"># Loading SCE object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">'data/data.Rdata'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span>,point.size <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Then run:</li>
</ol>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/rsconnect/" class="external-link">rsconnect</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">rsconnect</span><span class="fu">::</span><span class="fu">setAccountInfo</span><span class="op">(</span>name<span class="op">=</span><span class="st">"&lt;ACCOUNT&gt;"</span>, </span>
<span>                          token<span class="op">=</span><span class="st">"&lt;TOKEN&gt;"</span>,</span>
<span>                          secret<span class="op">=</span><span class="st">"&lt;SECRET&gt;"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">rsconnect</span><span class="fu">::</span><span class="fu">deployApp</span><span class="op">(</span><span class="st">'path/to/your/app.R'</span><span class="op">)</span></span></code></pre></div>
<p>Another option could be R Shiny Server, for more details please visit
the following <a href="https://posit.co/download/shiny-server/" class="external-link">link</a>. Finally, Amazon
Web Services (AWS) can also be used. A step-by-step tutorial on how a
shiny app can be deployed with AWS can be found <a href="https://www.charlesbordet.com/en/guide-shiny-aws/#" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="if-i-want-to-change-any-preprocessing-parameters-do-i-need-to-go-through-all-the-steps-again-to-relaunch-the-app">4. If I want to change any preprocessing parameters, do I need to go
through all the steps again to relaunch the app?<a class="anchor" aria-label="anchor" href="#if-i-want-to-change-any-preprocessing-parameters-do-i-need-to-go-through-all-the-steps-again-to-relaunch-the-app"></a>
</h2>
<p>No! createSCEobject() can take as input a SingleCellExpression object
previously normalized (logcounts assay) and with reduced dims already
estimated. In this case normalization and dimensionality reduction steps
could be avoided.</p>
<p>For instance, if you initiated the application by using test.type =
“wilcox” in paramFindMarkers but now intend to re-launch it utilizing a
different sort of test, like t-test, just use the createSCEobject()
function with the updated parameter, indicating that you do not require
the other features to be recalculated.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span></span>
<span>  xx <span class="op">=</span> <span class="va">scXample</span>,</span>
<span>  assay.name.raw <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  assay.name.normalization <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  partitionVars <span class="op">=</span> <span class="st">"inferred_cell_type"</span>, </span>
<span>  metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"source_name"</span>, <span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"strain"</span>, <span class="st">"treatment"</span>, <span class="st">"pseudotime"</span><span class="op">)</span>,</span>
<span>  calcRedDim <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  paramFindMarkers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>test.type <span class="op">=</span> <span class="st">"t"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="can-i-remove-bad-quality-cells-within-the-app">5. Can I remove bad quality cells within the app?<a class="anchor" aria-label="anchor" href="#can-i-remove-bad-quality-cells-within-the-app"></a>
</h2>
<p>You cannot filter cells inside the app. Instead you can identify and
download the cells id selected with lasso or box tool in “Find new
markers” tab inside “Markers”. Then you can run these lines of code to
filter those cells from the SCE object and redo the preprocessing
steps.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Following the quick start guide example:</span></span>
<span><span class="co">#library(scX)</span></span>
<span><span class="co">#cseo &lt;- createSCEobject(xx = scXample, </span></span>
<span><span class="co">#                        partitionVars = "inferred_cell_type", </span></span>
<span><span class="co">#                        metadataVars = c("source_name", "age", "sex", "strain", "treatment", "pseudotime"),</span></span>
<span><span class="co">#                        descriptionText = "Quick Start Guide")</span></span>
<span><span class="co">#launch_scX(cseo)</span></span>
<span><span class="co">## Inside the app select some cells with lasso or box tool in "Find new markers" tab and click 'Download Selected Cells'</span></span>
<span></span>
<span><span class="co"># Load cells id:</span></span>
<span><span class="va">cells2filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"/path/to/downloaded/file.csv"</span><span class="op">)</span></span>
<span> </span>
<span><span class="co"># Subsetting cells from SCE object:</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">cseo</span><span class="op">$</span><span class="va">SCE</span><span class="op">[</span>,<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cseo</span><span class="op">$</span><span class="va">SCE</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cells2filter</span><span class="op">$</span><span class="va">Selected_Cells</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># We run the function to compute all calculations again with the new object. The input object now has "logcounts" assay. </span></span>
<span><span class="co"># In order to compute a new normalization we set 'assay.name.normalization' to empty string and it will overwrite it:</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span>xx <span class="op">=</span> <span class="va">sce</span>,</span>
<span>                        assay.name.normalization <span class="op">=</span> <span class="st">""</span>,</span>
<span>                        partitionVars <span class="op">=</span> <span class="st">"inferred_cell_type"</span>,</span>
<span>                        metadataVars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"source_name"</span>, <span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"strain"</span>, <span class="st">"treatment"</span>, <span class="st">"pseudotime"</span><span class="op">)</span>,</span>
<span>                        descriptionText <span class="op">=</span> <span class="st">"Quick Start Guide"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="can-i-use-a-list-of-marker-genes-to-analyze-within-the-scx-app">6. Can I use a list of marker genes to analyze within the scX
app?<a class="anchor" aria-label="anchor" href="#can-i-use-a-list-of-marker-genes-to-analyze-within-the-scx-app"></a>
</h2>
<p>Yes, you can pass a list of marker genes for every cluster in any
partition of the data in the form of DataFrame through
<code>markerList</code>. The preprocess function will compute marker
statistics per gene for all clusters in the partitions passed through
<code>partitionVars</code> but it will subset the list of genes to those
present in the DataFrame.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Custom marker genes for some clusters</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>COP <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Neu4"</span>, <span class="st">"Bmp4"</span>, <span class="st">"Gpr17"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>NFOL1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Tcf7l2"</span>, <span class="st">"Frmd4a"</span>, <span class="st">"Pik3r3"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>NFOL2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fam107b"</span>, <span class="st">"Rras2"</span>, <span class="st">"Tmem2"</span>, <span class="st">"Cnksr3"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>MFOL1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Ctps"</span>, <span class="st">"Mag"</span>, <span class="st">"RIPTIDE"</span>)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>MFOL2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Wfdc18"</span>)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co"># data.frame preparation</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>partition <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"inferred_cell_type"</span>, <span class="at">times =</span> <span class="fu">length</span>(COP) <span class="sc">+</span> <span class="fu">length</span>(NFOL1) <span class="sc">+</span> <span class="fu">length</span>(NFOL2) <span class="sc">+</span> <span class="fu">length</span>(MFOL1) <span class="sc">+</span> <span class="fu">length</span>(MFOL2)))</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"COP"</span>, <span class="fu">length</span>(COP)), <span class="fu">rep</span>(<span class="st">"NFOL1"</span>, <span class="fu">length</span>(NFOL1)), <span class="fu">rep</span>(<span class="st">"NFOL2"</span>, <span class="fu">length</span>(NFOL2)), <span class="fu">rep</span>(<span class="st">"MFOL1"</span>, <span class="fu">length</span>(MFOL1)), <span class="fu">rep</span>(<span class="st">"MFOL2"</span>, <span class="fu">length</span>(MFOL2)))</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>gene <span class="ot">&lt;-</span> <span class="fu">c</span>(COP, NFOL1, NFOL2, MFOL1, MFOL2)</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co"># data.frame creation</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>ml <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Partition =</span> partition, <span class="at">Cluster =</span> cluster, <span class="at">Gene =</span> gene)</span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="co"># Code Example:</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="fu">library</span>(scX)</span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a>cseo <span class="ot">&lt;-</span> <span class="fu">createSCEobject</span>(<span class="at">xx =</span> scXample, </span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a>                        <span class="at">partitionVars =</span> <span class="st">"inferred_cell_type"</span>, </span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a>                        <span class="at">metadataVars =</span> <span class="fu">c</span>(<span class="st">"source_name"</span>, <span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"strain"</span>, <span class="st">"treatment"</span>, <span class="st">"pseudotime"</span>),</span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a>                        <span class="at">markerList =</span> ml</span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a>                        <span class="at">descriptionText =</span> <span class="st">"Quick Start Guide"</span>)</span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="fu">launch_scX</span>(cseo)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="how-can-i-use-scx-docker-image-for-singlecell-analysis">7. How can I use scX docker image for singlecell analysis?<a class="anchor" aria-label="anchor" href="#how-can-i-use-scx-docker-image-for-singlecell-analysis"></a>
</h2>
<p>Firstly, you have to pull scX docker image from DockerHub:</p>
<pre class="console"><code>docker pull chernolabs/scx</code></pre>
<p>After that you will need an R script to run scX analysis and a
Dockerfile to build the container. Suppose you have your SCE object
saved in /path/to/host/directory/data/sce.rds and an R script called
analysis_scX.R as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Library</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://chernolabs.github.io/scX/">scX</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Loading SCE object</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/home/data/sce.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Analysis</span></span>
<span><span class="va">cseo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSCEobject.html">createSCEobject</a></span><span class="op">(</span></span>
<span>    xx <span class="op">=</span> <span class="va">sce</span>,</span>
<span>    descriptionText <span class="op">=</span> <span class="st">"Using scX with Docker!"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Launch app</span></span>
<span><span class="fu"><a href="../reference/launch_scX.html">launch_scX</a></span><span class="op">(</span><span class="va">cseo</span>, port <span class="op">=</span> <span class="fl">9192</span>, host <span class="op">=</span> <span class="st">"0.0.0.0"</span>, launch.browser <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We are setting the port 9192 (you can choose your favorite port) and
launch.browser = FALSE. In the same directory, you have to create a
Dockerfile similar to this one:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode dockerfile"><code class="sourceCode dockerfile"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="kw">FROM</span> msbeckel/scx:0.3.0</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="kw">RUN</span> <span class="fu">mkdir</span> /home/shiny-app</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="kw">COPY</span> scx_analysis.R /home/shiny-app/make_scX.R</span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="kw">RUN</span> <span class="fu">chmod</span> <span class="at">-R</span> 755 /home/shiny-app/</span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="kw">RUN</span> <span class="fu">mkdir</span> /data <span class="kw">&amp;&amp;</span> <span class="fu">chown</span> shiny:shiny /data</span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a><span class="kw">EXPOSE</span> 9192</span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a><span class="kw">CMD</span> <span class="ex">Rscript</span> /home/shiny-app/scx_analysis.R</span></code></pre></div>
<p>With the Dockerfile complete, image is built with:</p>
<pre class="console"><code>docker build -t scx_example ./path/to/Dockerfile/directory</code></pre>
<p>and can be run as a standalone container with:</p>
<pre class="console"><code>docker run --name scx_example_app \
    -v /path/to/host/directory/data:/home/data \
    -p 9192:9192 \
    scx_example</code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/tvegawaichman" class="external-link">Tomás Vega Waichman</a>, <a href="https://github.com/mluzvercesi" class="external-link">María Luz Vercesi</a>, <a href="https://github.com/ariberar" class="external-link">Ariel A. Berardino</a>, <a href="https://github.com/msbeckel" class="external-link">Maximiliano S. Beckel</a>, Chernomoretz Lab and Collaborators.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
