<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="createSCEobject() returns a list which includes a SingleCellExperiment object with counts, at
least one clusterization, gene markers for each clusterization, reduced dimensions for
visualization, and any additional data provided. This list is used as input when launching
the scXplorer app."><title>SingleCellExperiment object ready for use in scXplorer — createSCEobject • scX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Roboto-0.4.1/font.css" rel="stylesheet"><link href="../deps/_JetBrains%20Mono-0.4.1/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SingleCellExperiment object ready for use in scXplorer — createSCEobject"><meta property="og:description" content="createSCEobject() returns a list which includes a SingleCellExperiment object with counts, at
least one clusterization, gene markers for each clusterization, reduced dimensions for
visualization, and any additional data provided. This list is used as input when launching
the scXplorer app."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scX</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/get_started.html">Quick Start</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tabs">Tabs</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tabs">
    <a class="dropdown-item" href="../articles/Summary.html">Summary</a>
    <a class="dropdown-item" href="../articles/Markers.html">Markers</a>
    <a class="dropdown-item" href="../articles/Gene_Expression.html">Gene Expression</a>
    <a class="dropdown-item" href="../articles/Differential_Expression.html">Differential Gene Expression</a>
    <a class="dropdown-item" href="../articles/Exploratory_Data_Analysis.html">Explotarory Data Analysis</a>
    <a class="dropdown-item" href="../articles/Visual_Tools.html">Visual Tools</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://www.leloir.org.ar/biologia-de-sistemas-integrativa?area=bioinformatica-y-biologia-computacional">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/tvegawaichman/scX">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>SingleCellExperiment object ready for use in scXplorer</h1>
      
      <div class="d-none name"><code>createSCEobject.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>createSCEobject()</code> returns a list which includes a SingleCellExperiment object with counts, at
least one clusterization, gene markers for each clusterization, reduced dimensions for
visualization, and any additional data provided. This list is used as input when launching
the scXplorer app.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">createSCEobject</span><span class="op">(</span></span>
<span>  <span class="va">xx</span>,</span>
<span>  assay.name.raw <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  assay.name.normalization <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  toFactors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  toKeep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chosen.hvg <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nHVGs <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  nPCs <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  calcRedDim <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  paramFindMarkers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>test.type <span class="op">=</span> <span class="st">"wilcox"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"any"</span><span class="op">)</span>,</span>
<span>  calcAllToFactors <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cells2keep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  descriptionText <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>xx</dt>
<dd><p>Either a matrix with counts, or a SCE or Seurat object.</p></dd>


<dt>assay.name.raw</dt>
<dd><p>Assay name for raw counts matrix if object is a SCE. Defaults to <code>counts</code>.</p></dd>


<dt>assay.name.normalization</dt>
<dd><p>Assay name for normalized matrix if present in SCE. If not present,
it computes <code>logcounts</code> (default).</p></dd>


<dt>metadata</dt>
<dd><p>Optional dataframe with cell metadata.</p></dd>


<dt>toFactors</dt>
<dd><p>Optional metadata column names (or <code>colData</code>) to use as factors. If <code>NULL</code> (default),
a quick clusterization will be computed.</p></dd>


<dt>toKeep</dt>
<dd><p>Additional metadata column names (or <code>colData</code>) to use only for coloring in plots.
If <code>NULL</code> (default), only <code>toFactors</code> columns will be available for coloring plots.</p></dd>


<dt>chosen.hvg</dt>
<dd><p>Optional list of Highly Variable Genes.</p></dd>


<dt>nHVGs</dt>
<dd><p>Number of Highly Variable Genes to use if <code>chosen.hvg=NULL</code>. Defaults to 3000.</p></dd>


<dt>nPCs</dt>
<dd><p>Number of Principal Components to use in PCA. Defaults to 50.</p></dd>


<dt>calcRedDim</dt>
<dd><p>Logical indicating whether to compute reduced dimensions (PCA, UMAP, TSNE, UMAP2D,
TSNE2D). Defaults to <code>TRUE</code>.</p></dd>


<dt>paramFindMarkers</dt>
<dd><p>List of parameters to pass to <code>scran::findMarkers(...)</code> to compute marker
genes for clusters. Defaults to <code>list(test.type="wilcox", pval.type="all", direction="any")</code>.</p></dd>


<dt>calcAllToFactors</dt>
<dd><p>Logical indicating whether to force the computation of markers and DEGs from
the entire list of <code>toFactors</code>.</p></dd>


<dt>cells2keep</dt>
<dd><p>List of names for cells to keep when subsampling data. NOTE: Subsampling is only
used in large datasets for visual purposes, and it does not affect computations.</p></dd>


<dt>descriptionText</dt>
<dd><p>Optional short description of the object being analized. This can help when
working with multiple tabs.</p></dd>


<dt>verbose</dt>
<dd><p>Logical for step by step status while function is running. Defaults to <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>List with a SingleCellExperiment object and additional data ready for use in <code>scXplorer</code>.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Tomás Vega Waichman, Maria Luz Vercesi, Ariel Agustín Berardino, Maximiliano Sebastian Beckel.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

